web/src/main/java/org/phoenixctms/ctsms/web/jersey/resource/NamedParameter.java

```diff
@@ -10,19 +12,21 @@ public NamedParameter(String name, Class<?> type) {
 		this.type = type;
 	}
 
+	@Override
+	public int hashCode() {
+		return Objects.hash(name);
```

Comment:
basically changes the behaviour. unless a dedciated hashCode() is implemented, it is ok/expected to use the hashCode() from supertype.

Reply:
This was brought up by sonarqube inspection: https://wiki.sei.cmu.edu/confluence/display/java/MET09-J.+Classes+that+define+an+equals%28%29+method+must+also+define+a+hashCode%28%29+method

Reply:
there are some (rather rare) cases in the code, where hashCode() is implemented explicitly, where it is needed to identify a object by it's content. in general this is not the cases, ie. even if two objects have same "name" field, it should be allowed to add them eg. twice to a set.

yes, i presumed that findbugs/sonartype etc. complain about such stuff by default.

Reply:
I'm not sure if you mean, that the implementation relies on a broken contract (e.g. `equals` and `hashcode` implementation disagree), or just that there are custom `hashcode()` and `equals()` implementations (which I'm completely fine with).

Please note that the implementation in this PR also just depends on `name`, but its using the Java 7 `Objects` class as helper utility.

Reply:
broken contract -> yes, like with Java Long comparison with values gt 127.

the general appraoch in code is that equals(), compareTo(), hashCode(), toString() ... are implemented (overloaded) only if there is explicit need for it. Here equals() was required and added, so it suffices (eg to sort or ad to map etc.) - w/o adding a dedicated hashCode().

hashCode() should reflect the literal object content. there are some definite places where equals() works in a tolerant way (eg. considers null and "" equal). these aspects got precedence over the contract argument.

Reply:
Java `Long` comparison gt 127 is not a broken contract, but just not very intuitive.
Whereas implementing equals but not hashcode is wrong. This is the contract: 

> If two objects are equal according to the equals(Object) method, then calling the hashcode() method on each of the two objects must produce the same integer result.

Reply:
i remember, it's an implication only, not equality. so this is legit 
- hashcode is equal but equals() returns false
seems that's why it got another sorted-out convention.

Reply:
looking at some code that uses NamedParameter, stuff changes if hashCode would be equal for instance with same "name" value.

```
	public LinkedHashSet<NamedParameter> getNamedParameters(String resource, boolean ...
	      LinkedHashMap<String, NamedParameter> namedParameters = new LinkedHashMap<String, NamedParameter>();
...
              return new LinkedHashSet<NamedParameter>(namedParameters.values());
	}
```

this code avoids hashCode() by constructing the set content explcitly beforehand, based on unique names. someone using contains() of the result will be safe -> default hashCode() is equal for identical object only, so it falls back to equals(), which mathes for equal name only.

Reply:
Later in the code: 
```java
...
Set<NamedParameter> queryParams = args.getNamedParameters(field.getName(), true);
if (addPsfQueryParams) {
	queryParams.addAll(PSFUriPart.SLURPED_NAMED_QUERY_PARAMETERS);
}
...
```
The manual checking of uniqueness fails at this piece.
But that wouldn't help because it's not about uniqueness of hashcode (which btw doesn't even need to be unique, it would be ok if hashcode would be constant for all objects, but the one thing thats problematic is if equals and hashcode disagree).

You might try out this piece of code to understand the implications of having hashcode be different from equals (for the second run, just comment the hashcode method):
```java
import java.util.ArrayList;
import java.util.LinkedHashSet;

public class NamedParam {

	public static void main(String[] args) {
		NamedParam x = new NamedParam("a",1);
		NamedParam y = new NamedParam("a",2);
		NamedParam z = new NamedParam("b",2);
		
		System.out.println(x.equals(y));
		ArrayList<NamedParam> in = new ArrayList<NamedParam>();
		
		in.add(x);
		in.add(y);
		in.add(z);
		
		LinkedHashSet<NamedParam> res = new LinkedHashSet<NamedParam>(in);
		System.out.println(res.size());
		System.out.println(res.contains(new NamedParam("a",1))); 
		System.out.println(res.contains(new NamedParam("a",2)));
		System.out.println(res.contains(new NamedParam("b",2)));
	}
	
	String a;
	int b;

	public NamedParam(String a, int b) {
		super();
		this.a = a;
		this.b = b;
	}
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((a == null) ? 0 : a.hashCode());
		return result;
	}
	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		NamedParam other = (NamedParam) obj;
		if (a == null) {
			if (other.a != null)
				return false;
		} else if (!a.equals(other.a))
			return false;
		return true;
	}
}
```

Reply:
i see, so this assumption "default hashCode() is equal for identical object only, so it falls back to equals(), which matches for equal name only" is wrong, it uses hashCode to shortcut actually.

so this change is actually a literal fix, convincing 👍 

to check if there is an overall impact i would compare the json response size of the /rest index.

Reply:
Output and ordering of content of __rest/__ changed due to swagger update, so it's not really possible to diff the changes caused by the fix. From a quick manual inspection it didn't change anything.

Reply:
yes, i therefor meant with and without chaning NamedParameters only

Reply:
Verified: Output of /rest didn't change after introducing hashcode for `NamedParameter`